docs/RFCs/0004-crashtracker-prevent-hangs.md

```diff
@@ -0,0 +1,81 @@
+
+1.  When the system is functioning normally, the crashtracker should, to the greatest extent possible, **have no impact on the functioning of a non-crashing system**.
+2.  When a crash does occur, the crashtracker should **prioritize the reliable collection and reporting of crash information**.
+3.  When a crash does occur, the crashtracker should **make a best-effort attempt to minimize impact** on the crashing process.
+
+In the [current architecture](https://github.com/DataDog/libdatadog/blob/main/docs/RFCs/0003-crashtracker-design.md#architectural-model), the crashtracker collector operates within the context of a crashing process.
+As a baseline requirement, crashtracker code should aim to be [signal safe](https://man7.org/linux/man-pages/man7/signal-safety.7.html).
+This is difficult, since some operations (e.g. stack walking), may do non signal safe operations such as `malloc`.
+Even if the code does avoid all non-signal-safe operations, the program itself may be in such a corrupted state that the crashtracker will itself crash or hang.
+
+In the context of a crashing customer process, a crash within the crashtracker is bad because it can disrupt existing debugging workflows.
+However, since the process was crashing anyway, it does not affect the overall availability of the service.
+A hang is worse, because it can prevent the process/pod from being restarted for a significant amount of time, potentially affecting the availability of the entire service.
+This RFC proposes an architectural mechanism to mitigate the risk of crashtracker hangs.
+
+## Proposed Architecture
+
+In the proposed architecture, the crashing process [clones](https://man7.org/linux/man-pages/man2/clone.2.html) a child process.
+Care should be taken during this process to avoid
+This child process (referred to as the **collector** in this RFC) is responsible for collecting the data required for crashtracking, while the parent operates as a watchdog (and is hence referred to as the **watchdog** process in the RFC).
+The **receiver** process remains as in RFC 3, either running as an [execve](https://man7.org/linux/man-pages/man2/execve.2.html) fork of the watchdog, or as an independent sidecar.
+If the child process completes successfully, the parent process reaps its PID, waits for the receiver process to finish uploading, and then chains the next signal handler, as it does currently.
+If the new child process, or the receiver process exceed their timeout budget, the watchdog [kills](https://man7.org/linux/man-pages/man2/kill.2.html) the offending child process, and then and then chains the next signal handler.
+
+### Timeouts
+
+Each component has a total timeout budget, which it must enforce.
+Since each component conducts a series of operations, the total budget will typically be subdivided among those operations, in a hierarchical fashion.
```

Comment:
Good call.  This actually implies a tension between
* Speed
* Emitting partial data to the agent/proxy/backend frequently

I believe we are correct to underspecify this at present, since the document only leaves open the possibility without creating an imperative.